### PR TITLE
Amd support, custom dwm build autologout, hacky runit support

### PR DESCRIPTION
**Pull request by Ann1kaB**
Reverted commit https://github.com/Shatur95/optimus-manager-qt/commit/bd6f6a8295b9324de4b7081fa3a867975b7441f9 since it doesn't seem to exist anywhere. 

Added custom dwm build install path for killProcess and a kinda hacky way to get runit to work as well as the AMD support stuff, which is basically reliant on the PR https://github.com/Askannz/optimus-manager/pull/268 getting merged.

Icons are broken due to the rename of Intel to integrated, briefly tested it and it looks like it would have worked if I wasn't using my optimus-manager fork because I use runit.

### CODE REVIEW

**Comment by Shatur** [comment]
Currently in this PR you use Intel icon for the integrated GPU. This will be a bit confusing for AMD users. I think that we should find an icon that would represent the integrated video card (or contains both AMD and Intel logos). What do you think about it? Maybe do you know any iconpack that have such icon?

**Comment by Ann1kaB** [comment]
> Currently in this PR you use Intel icon for the integrated GPU. This will be a bit confusing for AMD users. I think that we should find an icon that would represent the integrated video card (or contains both AMD and Intel logos). What do you think about it? Maybe do you know any iconpack that have such icon?

Actually it's funny you mention that, I do have one haha.
amd-intel

**Comment by Shatur** [comment]
:D

What if made a request to some iconpacks and just use one of them?

**Comment by Shatur** [comment]
Is configuration from the Intel settings page are applied to the AMD cards too?

**Comment by dphaldes** [comment]
I installed this from AUR but it stops in "detectGPU()". I looked through code and found this : 
image
My amdgpu is not at that address (if the term is correct)

```
xrandr --listproviders 
Providers: number : 2
Provider 0: id: 0x54 cap: 0xf, Source Output, Sink Output, Source Offload, Sink Offload crtcs: 4 outputs: 1 associated providers: 1 name:Unknown AMD Radeon GPU @ pci:0000:05:00.0
Provider 1: id: 0x1f7 cap: 0x2, Sink Output crtcs: 4 outputs: 1 associated providers: 1 name:NVIDIA-G0

```

**Comment by Shatur** [comment]
@luiq54, I think that we should check if string begins with.

**Comment by Shatur** [comment]
Done. Could you try it again?

**Comment by dphaldes** [comment]
Yay! Now it launches properly. I tried switching to Nvidia only and AMD only and it works properly but when switching to Hybrid, Optimus manager errors out 

switch log:
```
15] INFO: # Xorg pre-start hook
[15] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'integrated', 'current_mode': 'nvidia'}
[15] INFO: Requested mode is: integrated
[16] INFO: Checking for GDM display servers
[1237] INFO: Available modules: ['nouveau', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[1237] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
[1295] ERROR: Xorg pre-start setup error
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/bash.py", line 11, in exec_bash
    out = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['bash', '-c', 'modprobe -r nvidia_drm nvidia_modeset nvidia_uvm nvidia']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 197, in _unload_modules
    exec_bash("modprobe -r " + " ".join(modules_to_unload))
  File "/usr/lib/python3.8/site-packages/optimus_manager/bash.py", line 18, in exec_bash
    raise BashError(
optimus_manager.bash.BashError: Failed to execute 'modprobe -r nvidia_drm nvidia_modeset nvidia_uvm nvidia' :
modprobe: FATAL: Module nvidia is in use.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/hooks/pre_xorg_start.py", line 45, in main
    setup_kernel_state(config, prev_state, requested_mode)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 24, in setup_kernel_state
    _nvidia_down(config)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 65, in _nvidia_down
    _unload_nvidia_modules(available_modules)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 149, in _unload_nvidia_modules
    _unload_modules(available_modules, ["nvidia_drm", "nvidia_modeset", "nvidia_uvm", "nvidia"])
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 199, in _unload_modules
    raise KernelSetupError("Cannot unload modules %s : %s" % (str(modules_to_unload), str(e)))
optimus_manager.kernel.KernelSetupError: Cannot unload modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] : Failed to execute 'modprobe -r nvidia_drm nvidia_modeset nvidia_uvm nvidia' :
modprobe: FATAL: Module nvidia is in use.

[1297] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
[1297] INFO: Writing state {'type': 'pre_xorg_start_failed', 'switch_id': '20200828T150642', 'requested_mode': 'integrated'}
```

This might be due to me using custom version of Optimus Manger but I have only altered the file which it exports for hybrid mode. 
[My fork](https://github.com/luiq54/optimus-manager/tree/amd_support_fix)


EDIT: I think it error-ed  out when switching to integrated and not hybrid.

**Comment by Shatur** [comment]
Yes, it looks like the issue on the Optimus Manager side. This program is just a client that detects GPU, allows to edit settings via GUI and sends switch commands to the daemon socket. All switching operations done on the Optimus Manager side.

Have you tried https://github.com/Askannz/optimus-manager/pull/268? What are the differences between your fork?

**Comment by dphaldes** [comment]
My fork is based the `amd_support` branch on Askannz/optimus-manager. I had issues with switching to hybrid mode so I just patched the lines according to my [old xorg conf](https://pastebin.com/GYGUyfHa). I have just replaced some lines where it generates the xorg conf files. I haven't touched any code and manually switching from terminal works properly.

**Comment by Shatur** [comment]
Interesting, switch works from terminal, but not works from GUI, correct?

**Comment by dphaldes** [comment]
Yes. It works fine when using Terminal. Is there a way I can provide logs or something to help you figure out what the issue ?

**Comment by Shatur** [comment]
Hmm... I think I know what the problem is. In CLI client (default `optimus-manager` executable) you passing options to switch GPU. This options then passed to the daemon and it performs switching. What string does the Optimus Manager Daemon expect to switch the graphics card in your fork? It has previously accepted `hybrid`, `nvidia`, `intel` and `auto`.

**Comment by dphaldes** [comment]
I used `optimus-manger --help` and it displays following modes : 
```
--switch MODE         Set the GPU mode to MODE. You need to log out then log in to apply the change.Possible modes :
                        integrated, nvidia, hybrid, intel (deprecated, equivalent to integrated)
```

**Comment by Shatur** [comment]
Thanks for the information. When you switch, Optimus Manager Qt simply sends the `hybrid` string to the socket.
So, you should check if your fork handles switching correctly. This is probably should done on your side, I am poor in Python :)

**Comment by Shatur** [comment]
AMD support is now available in Optimus Manager master. Can anyone help me support it here?

File: src/optimussettings.h [hunk 1]
```diff
@@ -115,26 +115,26 @@ class OptimusSettings : public QSettings
...
     static GPU defaultExternalPowerStartupMode();
 
-    // Intel
-    Driver intelDriver() const;
-    void setIntelDriver(Driver driver);
-    static Driver defaultIntelDriver();
+    // Integrated
+    Driver integratedDriver() const;
+    void setIntegratedDriver(Driver driver);
+    static Driver defaultIntegratedDriver();
```
Comment (Shatur) [inline src/optimussettings.h]: If I understand correctly, AMD, like Intel, has separate settings. So, we need to add separate settings for AMD. I think we should just add the Intel and AMD prefixes to the enums.

File: src/settingsdialog.ui [hunk 2]
```diff
@@ -83,10 +83,10 @@
       </item>
       <item>
        <property name="text">
-        <string notr="true">Intel</string>
+        <string notr="true">Integrated</string>
```
Comment (Shatur) [inline src/settingsdialog.ui]: We should undo this and add a separate section for AMD.

File: CMakeLists.txt [hunk 3]
```diff
@@ -71,43 +71,43 @@ install(FILES ${QM_FILES} DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/${ORGANIZATIO
 install(FILES data/${DESKTOP_FILE} DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/applications)
 
 ecm_install_icons(ICONS
-    data/icons/freedesktop/128-apps-intel.svg
+    data/icons/freedesktop/128-apps-integrated.svg
```
Comment (Shatur) [inline CMakeLists.txt]: I think we should just use another icon for AMD.

**Review by Shatur — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (xacidr4inx) [reply CMakeLists.txt]: > I think we should just use another icon for AMD.

Yeah I agree

Comment (Shatur) [reply CMakeLists.txt]: It just should be taken from [this](https://github.com/masalla-art/masalla-icon-theme) theme (as Intel and Nvidia icons).

**Comment by Shatur** [comment]
I will continue to work on this in the master branch.
For future PRs, it'd be best to split independent changes into individual PRs.

**Pull request merged**